src/radical/pilot/pilot_manager.py

```diff
@@ -452,27 +452,30 @@ def _pilot_staging_input(self, pilot, directives):
 
     # --------------------------------------------------------------------------
     #
-    def _pilot_staging_output(self, pilot, directives):
+    def _pilot_staging_output(self, sds):
         '''
-        Run some staging directives for a pilot.  We pass this request on to
-        the launcher, and wait until the launcher confirms completion on the
-        command pubsub.
+        Run some staging directives for a pilot.
         '''
 
         # add uid, ensure its a list, general cleanup
-        sds  = expand_staging_directives(directives)
+        sds  = expand_staging_directives(sds)
         uids = [sd['uid'] for sd in sds]
-        self._active_sds = dict()
 
-        self.publish(rpc.CONTROL_PUBSUB, {'cmd': 'pilot_staging_output_request',
-                                          'arg': {'pilot' : pilot,
-                                                  'sds'   : sds}})
-        # keep track of SDS we sent off
-        for sd in sds:
-            sd['state'] = rps.NEW
-            self._active_sds[sd['uid']] = sd
+        # prepare to wait for completion
+        with self._sds_lock:
 
-        # and wait for their completion
+            self._active_sds = dict()
+            for sd in sds:
+                sd['state'] = rps.NEW
+                self._active_sds[sd['uid']] = sd
+
+            sd_states = [sd['state'] for sd
+                                     in  self._active_sds.values()
+                                     if  sd['uid'] in uids]
+        # push them out
+        self._stager_queue.put(sds)
+
+        # wait for their completion
         with self._sds_lock:
             sd_states = [sd['state'] for sd in self._active_sds.values()
                                             if sd['uid'] in uids]
```

Comment:
Is these 3 lines are needed? they kinda duplicate L472-474 even it regenerates to get updated states (it is similar to how this is organized in method `_pilot_staging_input`)

Reply:
You are right, removed one set of those lines.